Add attributes support

rustamwin opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️/❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any

ℹ️ https://github.com/yiisoft/demo/pull/589


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * PHP Attributes for declaring routes (Get, Post, Put, Delete, Patch, Head, Options, Route).
  * RoutesProvider interface plus array and file providers.
  * Routes/groups may be configured via constructors as well as fluent API.

* **Bug Fixes**
  * Aligned Group hosts behavior with Route hosts.
  * Constructor input validation added for hosts and middleware (invalid inputs now throw).

* **Documentation**
  * README updated to mention PHP Attributes support.

* **Tests**
  * Extensive tests added for attributes, providers, route/group behavior, and file fixtures.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rustamwin committed:
Add attributes support

StyleCIBot committed:
Apply fixes from StyleCI

vjik commented on src/MatchingResult.php:
```diff
@@ -86,6 +72,9 @@ public function methods(): array
     }
 
+    /**
+     * @psalm-assert-if-true true $this->isSuccess()
```
```suggestion
     * @psalm-assert-if-true true $this->isSuccess()
```

Why it? `psalm-assert-if-true` use for methods that boolean returns.

rustamwin committed:
Fix psalm annotation

rustamwin committed:
Add tests

xepozz committed:
Merge branch 'master' into add-attributes

xepozz committed:
Merge branch 'master' into add-attributes

rustamwin committed:
Add attributes registrar

StyleCIBot committed:
Apply fixes from StyleCI

rustamwin committed:
Add tests

rustamwin committed:
Fix tests & psalm issues

StyleCIBot committed:
Apply fixes from StyleCI

rustamwin committed:
Add test cases

StyleCIBot committed:
Apply fixes from StyleCI

vjik commented on src/RouteAttributesRegistrarInterface.php:
```diff
@@ -0,0 +1,18 @@
+ *
+ * @see https://www.php.net/manual/en/language.attributes.php
+ */
+interface RouteAttributesRegistrarInterface
```
Are you sure that interface needed?

rustamwin replied:
Yes, I've added an interface so that everyone can implement their own registrar routes via attributes.

vjik replied:
Where will be used this interface?

rustamwin replied:
https://github.com/yiisoft/demo/pull/589

rustamwin commented on src/Route.php:
```diff
@@ -295,31 +300,64 @@ public function __debugInfo()
+    /**
+     * @return array[]|callable[]|string[]
+     */
+    public function getBuiltMiddlewares(): array
```
Keep the current syntax is a must.

rustamwin replied:
Done.

rustamwin commented on src/Group.php:
```diff
@@ -4,53 +4,76 @@
 
 use function in_array;
 
+#[Attribute(Attribute::TARGET_CLASS)]
```
Yes.

samdark commented on src/RouteAttributesRegistrar.php:
```diff
@@ -0,0 +1,61 @@
+     */
+    public function register(): void
+    {
+        //TODO: caching?
```
?

rustamwin replied:
Reflection caching

samdark replied:
Either we need to remove that TODO and make it an issue in this repo or we need to implement that in this pull request.

rustamwin commented on src/Attribute/Delete.php:
```diff
@@ -0,0 +1,41 @@
+use Yiisoft\Http\Method;
+use Yiisoft\Router\Route;
+
+#[Attribute(Attribute::TARGET_METHOD | Attribute::IS_REPEATABLE)]
```
Adding attributes to `__invoke` method isn't enough?

xepozz replied:
In class scope it looks better

rustamwin replied:
Done.

rustamwin committed:
Improve Route

rustamwin committed:
Add test

xepozz committed:
Merge branch 'master' into add-attributes

rustamwin committed:
Increase coverage

rustamwin committed:
Merge remote-tracking branch 'origin/add-attributes' into add-attributes

xepozz committed:
Merge branch 'master' into add-attributes

rustamwin committed:
Add routes resource & attributes (#220)

* Add routes via resource + attributes

* Apply fixes from StyleCI

* Apply Rector changes (CI)

* Fix psalm

* Add missing file

* Apply fixes from StyleCI

* Rename Resource to Provider, add more tests

* Apply fixes from StyleCI

* Increase coverage

* Apply fixes from StyleCI

* Use variadic

---------

Co-authored-by: StyleCI Bot <bot@styleci.io>
Co-authored-by: rustamwin <rustamwin@users.noreply.github.com>

rustamwin committed:
Fix AttributeRoutesProvider

rustamwin committed:
Allow attributes to use in classes

rustamwin committed:
Fix psalm annotation

xepozz commented on tests/RouteTest.php:
```diff
@@ -420,6 +390,17 @@ public function testImmutability(): void
+                      ->middleware(TestMiddleware1::class)
+                      ->action(static fn () => new Response(200));
+
+        $builtMiddlewares = $route->getData('builtMiddlewares');
```
Due to the PR is going to raise the major version of package let's remove `getData` to regular getter?

vjik replied:
> Due to the PR is going to raise the major version of package let's remove `getData` to regular getter?

Rename `getData()`  to `get()` or create separate getters?

xepozz replied:
The second option

vjik replied:
This is very non-comfortable when configre routes. IDE suggestions will get in the way in application, but getters in most cases need into packages only.

xepozz commented on src/Group.php:
```diff
@@ -4,53 +4,63 @@
-        MiddlewareDispatcher $dispatcher = null
-    ): self {
-        return new self($prefix, $dispatcher);
+    public static function create(?string $prefix = null): self
```
How about to eliminate this?

vjik replied:
Create group via `new Group` is non-comfortable in application configuration.

vjik replied:
Most case of usage groups required chain calls.

1) Now:  `Group::create('/xxx')->...`
2) Your suggestion: `(new Group('/xxx'))->...`

Brackets in second way is non-comfortable =)

xepozz commented on src/MatchingResult.php:
```diff
@@ -86,6 +72,9 @@ public function methods(): array
     }
 
+    /**
+     * @psalm-assert-if-true !null $this->route
```
Why is it needed?

rustamwin replied:
Without assertion `PossiblyNullReference` psalm error occurs.

xepozz replied:
It shouldn't because there is check for null in the function body 😕

vjik commented on src/Route.php:
```diff
@@ -4,109 +4,119 @@
-
-    private array $disabledMiddlewareDefinitions = [];
-
+    private array $middlewares = [];
```
Keep previous name `middlewareDefinitions`. It more accurately reflects the contents of the property.

vjik commented on src/Route.php:
```diff
@@ -4,109 +4,119 @@
-    private bool $override = false;
-    private bool $actionAdded = false;
-
+    private array $builtMiddlewares = [];
```
This is also definitions, it is not built middlewares. Need rename.

vjik committed:
Add `RouteAttributeInterface` (#221)

rustamwin committed:
Adjust naming

rustamwin committed:
Minor improvements

rustamwin committed:
Minor

samdark commented:
TODO: move collector to https://github.com/yiisoft/router-composer-attribute-collector

rustamwin committed:
Move attribute collector it's own package

rustamwin committed:
Cleanup

rustamwin committed:
Minor

rustamwin committed:
Add changelog

rustamwin committed:
Apply Rector changes (CI)

rustamwin committed:
Merge remote-tracking branch 'origin/master' into add-attributes

# Conflicts:
#	.phpstorm.meta.php/Group.php
#	.phpstorm.meta.php/Route.php
#	src/Group.php
#	src/MatchingResult.php
#	src/Middleware/Router.php
#	src/Route.php
#	tests/GroupTest.php
#	tests/RouteCollectionTest.php
#	tests/RouteTest.php

rustamwin committed:
Fixes

rustamwin committed:
Apply Rector changes (CI)

StyleCIBot committed:
Apply fixes from StyleCI

rustamwin committed:
Fix psalm issues

vjik commented:
Waiting https://github.com/yiisoft/router/pull/225

rustamwin committed:
Merge remote-tracking branch 'origin/master' into add-attributes

# Conflicts:
#	composer.json
#	src/Group.php
#	src/Route.php
#	tests/RouteCollectionTest.php
#	tests/RouteTest.php

rustamwin committed:
Apply PHP CS Fixer and Rector changes (CI)

rustamwin committed:
Merge remote-tracking branch 'origin/master' into add-attributes

rustamwin committed:
Fix: BC

rustamwin committed:
Add tests for default override and middleware behavior; enhance file route provider handling; increase MSI

rustamwin committed:
Refactor: remove unused properties in Group; add return type to middleware callable in GroupTest

samdark committed:
Merge branch 'master' into add-attributes

rustamwin committed:
Refactor: extract host normalization logic to a shared private method; improve handling of route providers and test cases

rustamwin committed:
Apply PHP CS Fixer and Rector changes (CI)

samdark committed:
Apply PHP CS Fixer and Rector changes (CI)

rustamwin committed:
Update CHANGELOG: consolidate duplicate entries for #196

samdark commented on src/Attribute/Delete.php:
```diff
@@ -0,0 +1,54 @@
+     * @param bool $override Marks route as override. When added it will replace existing route with the same name.
+     * @param array $disabledMiddlewares Excludes middleware from being invoked when action is handled.
+     * It is useful to avoid invoking one of the parent group middleware for
+     * a certain route.
```
Let's reorder these annotations so these match the arguments' order. Same for all Attribute constructors.

rustamwin replied:
Good catch! Fixed.

samdark commented on src/Attribute/RouteAttributeInterface.php:
```diff
@@ -0,0 +1,12 @@
+
+use Yiisoft\Router\Route;
+
+interface RouteAttributeInterface
```
Is it used anywhere except in tests?

rustamwin replied:
It's used in https://github.com/yiisoft/router-composer-attribute-collector

samdark commented on src/Provider/FileRoutesProvider.php:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * @psalm-assert-if-true Route[]|Group[] $routes
+     */
+    private function isRoutesAreValid(array $routes): bool
```
```suggestion
    private function areRoutesValid(array $routes): bool
```

samdark commented on src/Provider/RoutesProviderInterface.php:
```diff
@@ -0,0 +1,19 @@
+use Yiisoft\Router\Route;
+
+/**
+ * `RoutesProviderInterface` provides routes.
```
```suggestion
 * `RoutesProviderInterface` provides routes to route collector.
```

samdark commented on src/Group.php:
```diff
@@ -39,9 +42,30 @@ final class Group
+     * @psalm-param list<array|callable|string> $middlewares
+     */
+    public function __construct(
         private ?string $prefix = null,
```
```suggestion
        private readonly ?string $prefix = null,
```

samdark commented on src/Group.php:
```diff
@@ -187,6 +204,32 @@ public function getData(string $key): mixed
         };
     }
 
+    private function assertHosts(array $hosts): void
```
```suggestion
    private function assertHostsValid(array $hosts): void
```

samdark commented on tests/Support/TestController.php:
```diff
@@ -7,11 +7,20 @@
+use Yiisoft\Router\Attribute\Get;
+use Yiisoft\Router\Group;
 
+#[Group('/test')]
```
Unused in tests?

samdark commented on tests/Support/TestController.php:
```diff
@@ -7,11 +7,20 @@
         return new Response(200, [], $request->getAttribute('content', ''));
     }
+
+    #[Get('/')]
```
Unused in tests?

samdark commented on src/Group.php:
```diff
@@ -124,14 +148,7 @@ public function host(string $host): self
-                $new->hosts[] = $host;
-            }
-        }
+        $new->hosts = $this->normalizeHosts($hosts);
```
The change corrects an error that was previously there. `Route::hosts()` was doing `$new->hosts = [];` but `Group::hosts()` was not. Now behavior is consistent. @rustamwin I think we need it in changelog as `Bug`.

rustamwin replied:
Done.

rustamwin committed:
Refactor: improve PHPDoc for route attributes and update validation method names in Group and FileRoutesProvider

rustamwin committed:
Apply PHP CS Fixer and Rector changes (CI)

rustamwin committed:
Update CHANGELOG: add bug fix entry for `Group::hosts()` method consistency with `Route::hosts()`

samdark committed:
Add phpdoc

diff --git c/src/Attribute/Delete.php i/src/Attribute/Delete.php
index 443a381..390034b 100644
--- c/src/Attribute/Delete.php
+++ i/src/Attribute/Delete.php
@@ -9,6 +9,9 @@ use Stringable;
 use Yiisoft\Http\Method;
 use Yiisoft\Router\Route;

+/**
+ * Route attribute that defines a DELETE HTTP method route.
+ */
 #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
 final class Delete implements RouteAttributeInterface
 {
diff --git c/src/Attribute/Get.php i/src/Attribute/Get.php
index 8fe2ed8..3d5ed51 100644
--- c/src/Attribute/Get.php
+++ i/src/Attribute/Get.php
@@ -9,6 +9,9 @@ use Stringable;
 use Yiisoft\Http\Method;
 use Yiisoft\Router\Route;

+/**
+ * Route attribute that defines a GET HTTP method route.
+ */
 #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
 final class Get implements RouteAttributeInterface
 {
diff --git c/src/Attribute/Head.php i/src/Attribute/Head.php
index 834b87a..ac78fe1 100644
--- c/src/Attribute/Head.php
+++ i/src/Attribute/Head.php
@@ -9,6 +9,9 @@ use Stringable;
 use Yiisoft\Http\Method;
 use Yiisoft\Router\Route;

+/**
+ * Route attribute that defines a HEAD HTTP method route.
+ */
 #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
 final class Head implements RouteAttributeInterface
 {
diff --git c/src/Attribute/Options.php i/src/Attribute/Options.php
index e62c71f..106a851 100644
--- c/src/Attribute/Options.php
+++ i/src/Attribute/Options.php
@@ -9,6 +9,9 @@ use Stringable;
 use Yiisoft\Http\Method;
 use Yiisoft\Router\Route;

+/**
+ * Route attribute that defines an OPTIONS HTTP method route.
+ */
 #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
 final class Options implements RouteAttributeInterface
 {
diff --git c/src/Attribute/Patch.php i/src/Attribute/Patch.php
index 9af032c..c4e310a 100644
--- c/src/Attribute/Patch.php
+++ i/src/Attribute/Patch.php
@@ -9,6 +9,9 @@ use Stringable;
 use Yiisoft\Http\Method;
 use Yiisoft\Router\Route;

+/**
+ * Route attribute that defines a PATCH HTTP method route.
+ */
 #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
 final class Patch implements RouteAttributeInterface
 {
diff --git c/src/Attribute/Post.php i/src/Attribute/Post.php
index 8f8bda6..a8fa0cc 100644
--- c/src/Attribute/Post.php
+++ i/src/Attribute/Post.php
@@ -9,6 +9,9 @@ use Stringable;
 use Yiisoft\Http\Method;
 use Yiisoft\Router\Route;

+/**
+ * Route attribute that defines a POST HTTP method route.
+ */
 #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
 final class Post implements RouteAttributeInterface
 {
diff --git c/src/Attribute/Put.php i/src/Attribute/Put.php
index 88ea1e9..9e05bc8 100644
--- c/src/Attribute/Put.php
+++ i/src/Attribute/Put.php
@@ -9,6 +9,9 @@ use Stringable;
 use Yiisoft\Http\Method;
 use Yiisoft\Router\Route;

+/**
+ * Route attribute that defines a PUT HTTP method route.
+ */
 #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
 final class Put implements RouteAttributeInterface
 {
diff --git c/src/Attribute/Route.php i/src/Attribute/Route.php
index cccdb14..5826ff2 100644
--- c/src/Attribute/Route.php
+++ i/src/Attribute/Route.php
@@ -8,6 +8,9 @@ use Attribute;
 use Stringable;
 use Yiisoft\Router\Route as RouteObject;

+/**
+ * Route attribute that defines a route with custom HTTP methods.
+ */
 #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
 final class Route implements RouteAttributeInterface
 {
diff --git c/src/Attribute/RouteAttributeInterface.php i/src/Attribute/RouteAttributeInterface.php
index cecfa63..c643960 100644
--- c/src/Attribute/RouteAttributeInterface.php
+++ i/src/Attribute/RouteAttributeInterface.php
@@ -6,7 +6,15 @@ namespace Yiisoft\Router\Attribute;

 use Yiisoft\Router\Route;

+/**
+ * Interface for route attributes that can provide a route instance.
+ */
 interface RouteAttributeInterface
 {
+    /**
+     * Returns the route instance defined by this attribute.
+     *
+     * @return Route The route instance.
+     */
     public function getRoute(): Route;
 }
diff --git c/src/Group.php i/src/Group.php
index 7717bde..6d4ea8f 100644
--- c/src/Group.php
+++ i/src/Group.php
@@ -13,6 +13,9 @@ use function is_array;
 use function is_callable;
 use function is_string;

+/**
+ * Route group that allows organizing routes with common properties.
+ */
 #[Attribute(Attribute::TARGET_CLASS)]
 final class Group
 {
@@ -77,6 +80,12 @@ final class Group
         return new self($prefix);
     }

+    /**
+     * Sets the routes for this group.
+     *
+     * @param self|Route ...$routes Routes or sub-groups to include in this group.
+     * @return self New instance with the specified routes.
+     */
     public function routes(self|Route ...$routes): self
     {
         $new = clone $this;
@@ -133,6 +142,12 @@ final class Group
         return $new;
     }

+    /**
+     * Sets the name prefix for all routes in this group.
+     *
+     * @param string $namePrefix Prefix to prepend to route names.
+     * @return self New instance with the specified name prefix.
+     */
     public function namePrefix(string $namePrefix): self
     {
         $new = clone $this;
@@ -140,11 +155,23 @@ final class Group
         return $new;
     }

+    /**
+     * Adds a host requirement for all routes in this group.
+     *
+     * @param string $host Host name to match.
+     * @return self New instance with the specified host.
+     */
     public function host(string $host): self
     {
         return $this->hosts($host);
     }

+    /**
+     * Sets host requirements for all routes in this group.
+     *
+     * @param string ...$hosts Host names to match.
+     * @return self New instance with the specified hosts.
+     */
     public function hosts(string ...$hosts): self
     {
         $new = clone $this;
@@ -172,22 +199,12 @@ final class Group
     }

     /**
-     * @psalm-template T as string
+     * Returns group data by key.
      *
-     * @psalm-param T $key
-     *
-     * @psalm-return (
-     *   T is ('prefix'|'namePrefix'|'host') ? string|null :
-     *   (T is 'routes' ? Group[]|Route[] :
-     *     (T is 'hosts' ? array<array-key, string> :
-     *       (T is ('hasCorsMiddleware') ? bool :
-     *         (T is 'enabledMiddlewares' ? list<array|callable|string> :
-     *           (T is 'corsMiddleware' ? array|callable|string|null : mixed)
-     *         )
-     *       )
-     *     )
-     *   )
-     * )
+     * @param string $key Data key to retrieve (`prefix`, `namePrefix`, `host`, `hosts`, `corsMiddleware`, `routes`,
+     * `hasCorsMiddleware`, `enabledMiddlewares`).
+     * @return mixed The requested data.
+     * @throws InvalidArgumentException If the key is unknown.
      */
     public function getData(string $key): mixed
     {
diff --git c/src/MatchingResult.php i/src/MatchingResult.php
index 3014823..5dfe6e1 100644
--- c/src/MatchingResult.php
+++ i/src/MatchingResult.php
@@ -7,6 +7,9 @@ namespace Yiisoft\Router;
 use RuntimeException;
 use Yiisoft\Http\Method;

+/**
+ * Result of matching a request against routes.
+ */
 final class MatchingResult
 {
     /**
@@ -51,6 +54,11 @@ final class MatchingResult
         return $this->route !== null;
     }

+    /**
+     * Checks if the request method was not allowed for the matched route.
+     *
+     * @return bool True if the method was not allowed, false otherwise.
+     */
     public function isMethodFailure(): bool
     {
         return $this->route === null && $this->methods !== Method::ALL;
diff --git c/src/Middleware/Router.php i/src/Middleware/Router.php
index e613d9c..6348ce7 100644
--- c/src/Middleware/Router.php
+++ i/src/Middleware/Router.php
@@ -17,10 +17,20 @@ use Yiisoft\Middleware\Dispatcher\MiddlewareFactory;
 use Yiisoft\Router\CurrentRoute;
 use Yiisoft\Router\UrlMatcherInterface;

+/**
+ * Router middleware that matches the request to a route and dispatches to the matched route's middleware.
+ */
 final class Router implements MiddlewareInterface
 {
     private readonly MiddlewareDispatcher $dispatcher;

+    /**
+     * @param UrlMatcherInterface $matcher URL matcher to find matching routes.
+     * @param ResponseFactoryInterface $responseFactory Factory for creating responses.
+     * @param MiddlewareFactory $middlewareFactory Factory for creating middleware instances.
+     * @param CurrentRoute $currentRoute Current route container.
+     * @param EventDispatcherInterface|null $eventDispatcher Optional event dispatcher.
+     */
     public function __construct(
         private readonly UrlMatcherInterface $matcher,
         private readonly ResponseFactoryInterface $responseFactory,
diff --git c/src/Provider/ArrayRoutesProvider.php i/src/Provider/ArrayRoutesProvider.php
index 3c52d54..5fa5f94 100644
--- c/src/Provider/ArrayRoutesProvider.php
+++ i/src/Provider/ArrayRoutesProvider.php
@@ -7,6 +7,9 @@ namespace Yiisoft\Router\Provider;
 use Yiisoft\Router\Route;
 use Yiisoft\Router\Group;

+/**
+ * Routes provider that is initialized with an array of routes and groups.
+ */
 final class ArrayRoutesProvider implements RoutesProviderInterface
 {
     /**
diff --git c/src/Provider/RoutesProviderInterface.php i/src/Provider/RoutesProviderInterface.php
index 16b6198..230fa46 100644
--- c/src/Provider/RoutesProviderInterface.php
+++ i/src/Provider/RoutesProviderInterface.php
@@ -8,11 +8,13 @@ use Yiisoft\Router\Group;
 use Yiisoft\Router\Route;

 /**
- * `RoutesProviderInterface` provides routes to route collector.
+ * Provides routes and route groups to route collector.
  */
 interface RoutesProviderInterface
 {
     /**
+     * Returns an array of routes and/or route groups.
+     *
      * @return Group[]|Route[]
      */
     public function getRoutes(): array;
diff --git c/src/Route.php i/src/Route.php
index 5faa63f..6f82f26 100644
--- c/src/Route.php
+++ i/src/Route.php
@@ -85,6 +85,11 @@ final class Route implements Stringable
         }
     }

+    /**
+     * Returns a string representation of the route.
+     *
+     * @return string String representation including name (if set), methods, hosts, and pattern.
+     */
     public function __toString(): string
     {
         $result = $this->name === null
@@ -108,6 +113,11 @@ final class Route implements Stringable
         return $result;
     }

+    /**
+     * Returns debug information about the route.
+     *
+     * @return array Array with route properties for debugging.
+     */
     public function __debugInfo()
     {
         return [
@@ -124,36 +134,78 @@ final class Route implements Stringable
         ];
     }

+    /**
+     * Creates a GET route.
+     *
+     * @param string $pattern URL pattern.
+     * @return self New route instance.
+     */
     public static function get(string $pattern): self
     {
         return self::methods([Method::GET], $pattern);
     }

+    /**
+     * Creates a POST route.
+     *
+     * @param string $pattern URL pattern.
+     * @return self New route instance.
+     */
     public static function post(string $pattern): self
     {
         return self::methods([Method::POST], $pattern);
     }

+    /**
+     * Creates a PUT route.
+     *
+     * @param string $pattern URL pattern.
+     * @return self New route instance.
+     */
     public static function put(string $pattern): self
     {
         return self::methods([Method::PUT], $pattern);
     }

+    /**
+     * Creates a DELETE route.
+     *
+     * @param string $pattern URL pattern.
+     * @return self New route instance.
+     */
     public static function delete(string $pattern): self
     {
         return self::methods([Method::DELETE], $pattern);
     }

+    /**
+     * Creates a PATCH route.
+     *
+     * @param string $pattern URL pattern.
+     * @return self New route instance.
+     */
     public static function patch(string $pattern): self
     {
         return self::methods([Method::PATCH], $pattern);
     }

+    /**
+     * Creates a HEAD route.
+     *
+     * @param string $pattern URL pattern.
+     * @return self New route instance.
+     */
     public static function head(string $pattern): self
     {
         return self::methods([Method::HEAD], $pattern);
     }

+    /**
+     * Creates an OPTIONS route.
+     *
+     * @param string $pattern URL pattern.
+     * @return self New route instance.
+     */
     public static function options(string $pattern): self
     {
         return self::methods([Method::OPTIONS], $pattern);
@@ -167,6 +219,12 @@ final class Route implements Stringable
         return new self($methods, $pattern);
     }

+    /**
+     * Sets the route name.
+     *
+     * @param string $name Route name.
+     * @return self New instance with the specified name.
+     */
     public function name(string $name): self
     {
         $route = clone $this;
@@ -174,6 +232,12 @@ final class Route implements Stringable
         return $route;
     }

+    /**
+     * Sets the URL pattern.
+     *
+     * @param string $pattern URL pattern.
+     * @return self New instance with the specified pattern.
+     */
     public function pattern(string $pattern): self
     {
         $new = clone $this;
@@ -181,11 +245,23 @@ final class Route implements Stringable
         return $new;
     }

+    /**
+     * Adds a host requirement.
+     *
+     * @param string $host Host name to match.
+     * @return self New instance with the specified host.
+     */
     public function host(string $host): self
     {
         return $this->hosts($host);
     }

+    /**
+     * Sets host requirements.
+     *
+     * @param string ...$hosts Host names to match.
+     * @return self New instance with the specified hosts.
+     */
     public function hosts(string ...$hosts): self
     {
         $route = clone $this;
@@ -304,24 +380,12 @@ final class Route implements Stringable
     }

     /**
-     * @psalm-template T as string
+     * Returns route data by key.
      *
-     * @psalm-param T $key
-     *
-     * @psalm-return (
-     *   T is ('name'|'pattern') ? string :
-     *       (T is 'host' ? string|null :
-     *           (T is 'hosts' ? array<array-key, string> :
-     *               (T is 'methods' ? array<array-key,string> :
-     *                   (T is 'defaults' ? array<string,string> :
-     *                       (T is ('override'|'hasMiddlewares') ? bool :
-     *                           (T is 'enabledMiddlewares' ? array<array-key,array|callable|string> : mixed)
-     *                       )
-     *                   )
-     *               )
-     *           )
-     *       )
-     *    )
+     * @param string $key Data key to retrieve (`name`, `pattern`, `host`, `hosts`, `methods`, `defaults`, `override`,
+     * `hasMiddlewares`, `enabledMiddlewares`).
+     * @return mixed The requested data.
+     * @throws InvalidArgumentException If the key is unknown.
      */
     public function getData(string $key): mixed
     {
diff --git c/src/RouteCollection.php i/src/RouteCollection.php
index 5af3ef6..54bebe3 100644
--- c/src/RouteCollection.php
+++ i/src/RouteCollection.php
@@ -13,6 +13,8 @@ use function in_array;
 use function is_array;

 /**
+ * Collection of routes that manages route registration and builds a route tree.
+ *
  * @psalm-type Items = array<array-key,array|string>
  */
 final class RouteCollection implements RouteCollectionInterface
@@ -29,6 +31,9 @@ final class RouteCollection implements RouteCollectionInterface
      */
     private array $routes = [];

+    /**
+     * @param RouteCollectorInterface $collector The route collector to use.
+     */
     public function __construct(private readonly RouteCollectorInterface $collector) {}

     public function getRoutes(): array
diff --git c/src/RouteCollectionInterface.php i/src/RouteCollectionInterface.php
index fe2ff1f..107804d 100644
--- c/src/RouteCollectionInterface.php
+++ i/src/RouteCollectionInterface.php
@@ -4,17 +4,31 @@ declare(strict_types=1);

 namespace Yiisoft\Router;

+/**
+ * Interface for route collections that provide access to registered routes.
+ */
 interface RouteCollectionInterface
 {
     /**
-     * @return Route[]
+     * Returns all routes in the collection.
+     *
+     * @return Route[] Array of routes indexed by name.
      */
     public function getRoutes(): array;

+    /**
+     * Returns a route by name.
+     *
+     * @param string $name Route name.
+     * @return Route The route instance.
+     * @throws RouteNotFoundException If the route is not found.
+     */
     public function getRoute(string $name): Route;

     /**
      * Returns routes tree array.
+     *
+     * @return array Hierarchical array of routes and/or groups.
      */
     public function getRouteTree(): array;
 }
diff --git c/src/RouteCollector.php i/src/RouteCollector.php
index c44479b..7c4b959 100644
--- c/src/RouteCollector.php
+++ i/src/RouteCollector.php
@@ -6,6 +6,9 @@ namespace Yiisoft\Router;

 use Yiisoft\Router\Provider\RoutesProviderInterface;

+/**
+ * Simple route collector that manages routes, groups, and middleware definitions.
+ */
 final class RouteCollector implements RouteCollectorInterface
 {
     /**
@@ -25,6 +28,12 @@ final class RouteCollector implements RouteCollectorInterface
      */
     public function __construct(private readonly array $providers = []) {}

+    /**
+     * Adds routes or groups to the collector.
+     *
+     * @param Route|Group ...$routes Routes or groups to add.
+     * @return RouteCollectorInterface The collector instance.
+     */
     public function addRoute(Route|Group ...$routes): RouteCollectorInterface
     {
         array_push(
@@ -52,6 +61,11 @@ final class RouteCollector implements RouteCollectorInterface
         return $this;
     }

+    /**
+     * Returns all registered items (routes and groups).
+     *
+     * @return Group[]|Route[]
+     */
     public function getItems(): array
     {
         if (!$this->providersAreInjected) {
@@ -66,6 +80,11 @@ final class RouteCollector implements RouteCollectorInterface
         return $this->items;
     }

+    /**
+     * Returns all middleware definitions.
+     *
+     * @return array[]|callable[]|string[]
+     */
     public function getMiddlewareDefinitions(): array
     {
         return $this->middlewareDefinitions;
diff --git c/src/RouteCollectorInterface.php i/src/RouteCollectorInterface.php
index 38c5861..dca8742 100644
--- c/src/RouteCollectorInterface.php
+++ i/src/RouteCollectorInterface.php
@@ -4,6 +4,9 @@ declare(strict_types=1);

 namespace Yiisoft\Router;

+/**
+ * Interface for route collectors that manage route registration.
+ */
 interface RouteCollectorInterface
 {
     /**
@@ -14,21 +17,31 @@ interface RouteCollectorInterface
     /**
      * Appends a handler middleware definition that should be invoked for a matched route.
      * First added handler will be executed first.
+     *
+     * @param array|callable|string ...$middlewareDefinition Middleware definitions.
+     * @return self New instance with the middleware appended.
      */
     public function middleware(array|callable|string ...$middlewareDefinition): self;

     /**
      * Prepends a handler middleware definition that should be invoked for a matched route.
      * First added handler will be executed last.
+     *
+     * @param array|callable|string ...$middlewareDefinition Middleware definitions.
+     * @return self New instance with the middleware prepended.
      */
     public function prependMiddleware(array|callable|string ...$middlewareDefinition): self;

     /**
+     * Returns all registered items (routes and groups).
+     *
      * @return Group[]|Route[]
      */
     public function getItems(): array;

     /**
+     * Returns all middleware definitions.
+     *
      * @return array[]|callable[]|string[]
      */
     public function getMiddlewareDefinitions(): array;
diff --git c/src/UrlMatcherInterface.php i/src/UrlMatcherInterface.php
index f802236..efef6c0 100644
--- c/src/UrlMatcherInterface.php
+++ i/src/UrlMatcherInterface.php
@@ -7,10 +7,15 @@ namespace Yiisoft\Router;
 use Psr\Http\Message\ServerRequestInterface;

 /**
- * `UrlMatcherInterface` allows finding a matching route given a PSR-8 server request. It is preferred to type-hint
- * against it in case you need to match URL.
+ * `UrlMatcherInterface` allows finding a matching route given a server request.
  */
 interface UrlMatcherInterface
 {
+    /**
+     * Matches a server request against registered routes.
+     *
+     * @param ServerRequestInterface $request The server request to match.
+     * @return MatchingResult The result of matching, containing route and parameters if successful.
+     */
     public function match(ServerRequestInterface $request): MatchingResult;
 }

samdark commented:
For moving to constructor syntax for routes, methods should be marked as deperecated.

rustamwin committed:
Mark `Group::create`, `RouteCollectorInterface`, `RouteCollector`, and `Route` HTTP methods as deprecated (#287)

samdark committed:
Update src/Route.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

samdark committed:
Fix incorrectly backtick usage and wrong class name in @deperecated

samdark committed:
Add validation for single route case

samdark committed:
Apply PHP CS Fixer and Rector changes (CI)

samdark commented:
```php
<?php

declare(strict_types=1);

use App\Web;
use Yiisoft\Router\Group;
use Yiisoft\Router\Route;

return [
    Group::create()
        ->routes(
            Route::get('/')
                ->action(Web\HomePage\Action::class)
                ->name('home'),
            Route::get('/say[/{message}]')
                ->action(Web\Echo\Action::class)
                ->name('echo/say'),
        ),
];
```

becomes

```php
<?php

declare(strict_types=1);

use App\Web;
use Yiisoft\Router\Group;
use Yiisoft\Router\Route;
use Yiisoft\Http\Method;

return [
    Group::create()
        ->routes(
            new Route([Method::GET], '/')
                ->action(Web\HomePage\Action::class)
                ->name('home'),
            new Route([Method::GET], '/say[/{message}]')
                ->action(Web\Echo\Action::class)
                ->name('echo/say'),
        ),
];
```

I think that the syntax is a bit clumsy like that.

samdark commented on src/Route.php:
```diff
@@ -12,47 +12,84 @@
-    private function __construct(
-        private array $methods,
+    public function __construct(
+        array $methods,
```
```suggestion
        array|string $method,
```

Let's make it possible to use

```php
new Route('GET', ...
```

samdark commented on src/Route.php:
```diff
@@ -12,47 +12,84 @@
         private string $pattern,
-    ) {}
+        private ?string $name = null,
+        array|callable|string|null $action = null,
```
```suggestion
        array|callable|string|null $action = null,
        private ?string $name = null,
```

Action is used more than name.

samdark commented on src/Group.php:
```diff
@@ -39,20 +45,49 @@ final class Group
-        private ?string $prefix = null,
-    ) {}
+    /**
+     * @param array[]|callable[]|string[] $middlewares Middleware definitions.
```
Missing phpdoc.

samdark commented on src/Route.php:
```diff
@@ -12,47 +12,84 @@
      */
-    private function __construct(
-        private array $methods,
+    public function __construct(
```
Missing phpdoc.

samdark committed:
Allow passing method as a single value instead of an array

samdark committed:
Action is used more than name, swap order

samdark committed:
Add missing phpdoc

samdark committed:
Add missing Group phpdoc

samdark committed:
Remove unclear statement

samdark commented on README.md:
```diff
@@ -28,6 +28,7 @@ with an adapter package. Currently, the only adapter available is [FastRoute](ht
 - Ready to use middleware for route matching.
 - Convenient `CurrentRoute` service that holds information about last matched route.
 - Out of the box CORS middleware support.
+- Declaring routes using PHP attributes.
```
There should be an example for using attributes and then reading these with `yiisoft/router-composer-attribute-collector`.

samdark commented on src/Attribute/RouteAttributeInterface.php:
```diff
@@ -0,0 +1,20 @@
+use Yiisoft\Router\Route;
+
+/**
+ * Interface for route attributes that can provide a route instance.
```
Need to mention that attributes that implement this interface are read by `yiisoft/router-composer-attribute-collector` package.

samdark commented on src/Provider/ArrayRoutesProvider.php:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Routes provider that is initialized with an array of routes and groups.
+ */
+final class ArrayRoutesProvider implements RoutesProviderInterface
```
Is there a use-case besides tests?

rustamwin replied:
Don't remember exactly, maybe yes could be useful for 3rd-party library.

samdark committed:
Use cast instead of check to get an array from $method

samdark committed:
Apply PHP CS Fixer and Rector changes (CI)

samdark commented on src/Route.php:
```diff
@@ -279,24 +402,12 @@ public function disableMiddleware(mixed ...$definition): self
-     *               )
-     *           )
-     *       )
-     *    )
```
Look like Psalm type was useful.

rustamwin replied:
Reverted.

samdark commented on CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
+- New #196: Add PHP Attributes support (@rustamwin)
+- New #196: Add `RoutesProviderInterface` interface providing routes from various resources (@rustamwin)
+- Bug #196: Fix the behavior of `Group::hosts()` method to be consistent with `Route::hosts()` method (@rustamwin)
```
Need to mention the deprecation of `Route` static constructors and describe these in the README.

samdark committed:
Compare with empty array instead of using `empty()`

rustamwin committed:
Fix: suppress psalm warnings and enhance PHPDocs in Route and RouteCollector

rustamwin committed:
Apply PHP CS Fixer and Rector changes (CI)

rustamwin committed:
Fix: suppress psalm warnings and improve PHPDocs in RouteCollection and Group

rustamwin committed:
Merge remote-tracking branch 'origin/add-attributes' into add-attributes

rustamwin committed:
Fix: suppress psalm warnings and enhance PHPDocs in RouteCollection

rustamwin committed:
Fix: adjust psalm suppression placement in RouteCollection PHPDoc

rustamwin committed:
Kill mutation

rustamwin committed:
Refactor: mark `$container` as readonly in `RouterCollector` and update README with attributes usage example
